### ISSUE TITLE
Generate trie lm::FormatLoadException

### ISSUE THREAD

**Issue by yoann1995** [mozilla/DeepSpeech#1407]
I'm following this tutorial : https://discourse.mozilla.org/t/tutorial-how-i-trained-a-specific-french-model-to-control-my-robot/22830 to create a French model. 

The problem is when generating the trie file with this command : 

> ./generate_trie data/cassia/alphabet.txt data/cassia/lm.binary data/cassia/vocabulary.txt data/cassia/trie

I have this output : 

> terminate called after throwing an instance of 'lm::FormatLoadException'
  what():  native_client/kenlm/lm/binary_format.cc:131 in void lm::ngram::MatchCheck(lm::ngram::ModelType, unsigned int, const lm::ngram::Parameters&) threw FormatLoadException.
The binary file was built for probing hash tables but the inference code is trying to load trie with quantization and array-compressed pointers
Abandon (core dumped)

I tried several times to generate my lm.binary with kenlm (./build_binary -T -s words.arpa  lm.binary) but still the same error.

**Comment by hhk998402** [comment]
Facing the same issue on MacOS Sierra

**Comment by reuben** [comment]
We've very recently switched the format of our language model and tooling on master. It looks like you're using the pre-built generate_trie binary from the 0.1.1 release. You can fix this by using consistent versions of the various tools, including the training code. If you want to use the 0.1.1 binaries, work from the v0.1.1 tag, not the master branch. Alternatively you can work from master and build things yourself.

Hope this helps.

**Comment by yoann1995** [comment]
Thank you for your reply ! I tried to use the V.0.1.1 tag but I still have the same issue. 

If I want to stay on the master branch, which tools can I use to generate langage model with this new format ? (Because kenlm seems to use only probing hash tables).

**Comment by reuben** [comment]
KenLM does not use probing hash tables only: https://kheafield.com/code/kenlm/structures/

Depending on which structure you use for your KenLM model you'll need to adjust the typedefs in `beam_search.h` and `generate_trie.cpp`.

Build `generate_trie` following the normal commands in the [documentation](https://github.com/mozilla/DeepSpeech/blob/master/native_client/README.md#build-requirements).

**Comment by reuben** [comment]
> Thank you for your reply ! I tried to use the V.0.1.1 tag but I still have the same issue.

That should not happen if you're using `generate_trie` of version v0.1.1 as well.

**Comment by yoann1995** [comment]
I followed the step to build the generate_trie in the v0.1.1 version but I don't have any generated_trie file created at the end. I must miss something. I can only have this file by using the command : 

> python util/taskcluster.py --target . 
But if I understand well, this will give me the generate_trie file which use the quantization and array-compressed pointers format ?

**Comment by yoann1995** [comment]
No because the newer file format is : quantization and array-compressed pointers and I need probing hash tables. :/

**Comment by lissyx** [comment]
@yoann1995 Why don't you get the binary from v0.1.1 then ?

**Comment by lissyx** [comment]
https://index.taskcluster.net/v1/task/project.deepspeech.deepspeech.native_client.v0.1.1.cpu/artifacts/public/native_client.tar.xz

**Comment by yoann1995** [comment]
Well, I try to do it from 0.1.1 but I still have the same issue. Thank you for the link, it works now ! Sorry for the inconvenience and thank you for your help ! :)

**Comment by lissyx** [comment]
Still, it'd be good we figure out why master you cannot get that. Maybe we need to fix some docs ? We did the change rather soon, as @reuben said, so it's possible we lack something :), and I don't remember the flags for KenLM. Is it possible the `binary` being referred to is actually `lm.binary` and not the `generate_trie` binary ?

**Comment by yoann1995** [comment]
The generate_trie binary is not really the problem for me. The real problem is that I don't know how to generate my lm.binary file other than with the probing hash tables format.

**Comment by lissyx** [comment]
Yeah, which makes sense since Vincent's guide is old now :-). I should still have a kenlm build ready that I used to test that, I'll try and find the proper arguments :-)

**Comment by yoann1995** [comment]
Okay, thanks ! :)

**Comment by lissyx** [comment]
@yoann1995 
```
Usage: ./build_binary [-u log10_unknown_probability] [-s] [-i] [-w mmap|after] [-p probing_multiplier] [-T trie_temporary] [-S trie_building_mem] [-q bits] [-b bits] [-a bits] [type] input.arpa [output.mmap]
```

So make sure `[type]` is `trie` and not `probing`, I think this is what you need.

**Comment by hhk998402** [comment]
Thank you for your reply :)
This worked for me on Mac - https://index.taskcluster.net/v1/task/project.deepspeech.deepspeech.native_client.v0.1.1.osx/artifacts/public/native_client.tar.xz

This works for all other Linux systems - 
https://index.taskcluster.net/v1/task/project.deepspeech.deepspeech.native_client.v0.1.1.cpu/artifacts/public/native_client.tar.xz

**Comment by unicornmafia** [comment]
I am getting a similar error:

**Comment by spate141** [comment]
I'm having an issue with building trie model with generate_trie script! Only issue I'm having is; building process just get killed after using almost 40GB of memory! My binary language model is of size 1.8 GB, Vocab file is around 25GB. Is there any way to determine how much memory is required to generate trie file?

**Comment by lissyx** [comment]
@spate141 I'd rather question the size of your files, it looks like way too big.

**Comment by spate141** [comment]
@lissyx Thanks for the reply! Yes, the file size was the issue; so I created smaller version of language model`(ngram 1=2782949 ngram 2=2874575 ngram 3=1665316)` from a small vocab.txt file, and I was able to generate a new `trie` file along with it.

One more thing I noticed, If I use my newly trained language model with DeepSpeech's trie model file I'm getting more accurate transcriptions rather than If I use my custom lm with a newly generated trie model file. I'm not sure about this behavior. Any thoughts?

**Comment by lissyx** [comment]
@spate141 Can we take that discussion on Discourse ?

**Comment by spate141** [comment]
@lissyx: Adding in example from this video: https://www.youtube.com/watch?v=JRAdZyW_hss **(0:00 to 2:00)**. I have highlighted words that are different in both transcript. I'm using same [DeepSpeech 0.1.1](https://github.com/mozilla/DeepSpeech/releases/download/v0.1.1/deepspeech-0.1.1-models.tar.gz) acoustic model with different language model. 

> **New LM + DeepSpeech TRIE:**

he is its manda and say i'm knowing to show you have removed a red wine stain for my carpet i able did not think it would come out and it totally did and i'm super excited um because i just had talked it up and decided that the carpet was totally ruined and i just **go gold** some ideas on how to get rid one now and im going to kind of explain to you the steps that i did and i really impressed with **hot** work a so this is the red wine stain on the carpet and you can see it s very very purple our carpet is **super old and i to place if i can get a sound how can be** devastated but **im** all i have done at this point is last night i stole the wine and i just got a wet wash cloth and kind of dab and patted at it to soak up the excess and kinda let i get wet i did not clean it last night um because unfortunately if you happen enough fund that your spilling wine **on the carpet you re probably not in a cleaning up wine stain state ah so** that was my situation so i'm gonna try it today my first method is going to be with peroxide **and baking so i m** starting with bro feta baking so dt a because i just happen **have to of these on hand** and it was one of the solutions i read on line so i said to spray it with hide jem fer suf pro xian did on t have an empty spray bottles on is gonna condo por it and **saturate** it with an the hind rd jim pro x kid ed then it said to im kino spread the baking soda over top of it and let it sit for two to three minutes and then just clean it up so thats what i m going to **dry it try first** and we'll see the kind of difference it make so there s my before li tings naw the best in here hopefully i will have a better after a so here is my **proxide and being so a** spot **treatment** jus kana f sitting on n the **carpet** i really don't think this is going to do anything i thought i **it is or** **some thing** when i

> **NEW LM + NEW TRIE**

he is its manda and say i'm knowing to show you have removed a red wine stain for my carpet i able did not think it would come out and it totally did and i'm super excited um because i just had talked it up and decided that the carpet was totally ruined and i just **do good** some ideas on how to get rid one ow and i'm gonna kind of explain to you the steps that i did and i really impressed with **h out** work a so this is the red wine stain on the carpet and you can see it s very very purple our carpet **ids epr old inou io places if a can k get hso u and how t can be** devastated but **i mall** i have done at this point is last night i stole the wine and i just got a wet wash cloth and kind of dab and patted at it to soak up the excess and kinda let i get wet i did not clean it last night um because unfortunately if you happen enough fund that your spilling wine **on the gar pet you re probly not in a cle a ming up wy ne staind state a hso** that was my situation so i'm gonna try it today my first method is going to be with peroxide **amba kings odon** starting with bro feta baking so dt a because ti just happen t**ham to s of these on he nd** and it was one of the solutions i read on line so it said to spray it with hide jem fer suf pro xian did on t have an empty spray bottles on is gonna condo por it and **satu raed** it with om aid to im kina spread the baking so a over top of it n let its it for two to three minutes and then just clean it up so that s what i m going to **dry try first** and we'll see the kind of difference it make so there s my before w lit dings naw the best in here hopefully i will have a better after a so here is my **pr oxide an nd be a king soda** spot **treat men** jus kana f sitting on n the **par pet** i really don't think this is going to do anything i thought i **it d is z orr** **something** whan i

**Comment by spate141** [comment]
I'm using [v0.2.0-alpha.6](https://github.com/mozilla/DeepSpeech/releases/tag/v0.2.0-alpha.6) `native_client` binary files, and getting similar type of error:

My kenlm language model is based on trie; here command:
`./build_binary trie lm.arpa lm.binary`

Error on `generate_trie`:
```
./generate_trie alphabet.txt lm.binary vocab.txt trie

terminate called after throwing an instance of 'lm::FormatLoadException'
  what():  native_client/kenlm/lm/binary_format.cc:131 in void lm::ngram::MatchCheck(lm::ngram::ModelType, unsigned int, const lm::ngram::Parameters&) threw FormatLoadException.
The binary file was built for trie but the inference code is trying to load trie with quantization and array-compressed pointers
Aborted (core dumped)
```

**Comment by lissyx** [comment]
@spate141 Yes, the format changed, use the proper parameters to rebuild the language model, see my comment above.

**Comment by abuvaneswari** [comment]
@spate141 , this command for kenlm build_binary fixed the issue:
bin/build_binary trie -q 16 -b 7 -a 64 set5.arpa set5.lm.binary

**Comment by abuvaneswari** [comment]
@lissyx 
* Using the generate_trie of v0.2.0 alpha6
* using the alphabet.txt in DeepSpeech/data/alphabet.txt
* using an lm.binary (trie format) that I generated after following instructions above
* using vocab that is essentially the text source to the lm.binary generation. The num lines of my vocab.txt = 4 Million

$ ./generate_trie ../tensorflow/DeepSpeech/data/alphabet.txt lm.binary set5.txt /tmp/trie_lm_trie_long

terminate called after throwing an instance of 'boost::locale::conv::conversion_error'
  what():  Conversion failed
Aborted

However, if I use your vocab.txt (available in the DeepSpeech/data/lm/vocab.txt (Num lines = 94k), with my lm.binary, the generate_trie is able to generate a trie for me successfully.

I have observed the same behavior even with version 0.1.1 of your generate_trie (that is, when I supply a vocab = 4 Million lines, it would throw this error whereas when I supply your vocab, it would generate the trie). So, this is not an issue with generate_trie version.

My question to you:
* What essentially should go into the vocab file? What is the role played by the vocab file? Is n't the lm.binary alone enough to do the language model lookup?
* I believe the DeepSpeech/data/lm/vocab file consists of all transcripts from TED talks. Do you use this same file or a different one when you generate the trie for DS models official release?
* What do I lose if I use the DeepSpeech/data/lm/vocab.txt in the place of my own 4 Million line long vocab file to generate trie?

**Comment by lissyx** [comment]
@abuvaneswari Please avoid hijacking issues for questions like that, please use proper code formatting for including code / console output.

I have no idea why you have this `boost::locale` error, likely some bogus data somewhere. Since it works with our data, you need to find what is wrong on your side.

The file `vocab.txt` is used to build the language model `lm.binary`.

The official model release were using some `vocab.txt` file we were unable to redistribute ; not sure if it was TED talks. This is changing with 0.2.0.

You might loose accuracy and time debugging it by using another `vocab.txt`. We spent weeks training and comparing different dataset and build parameters for selecting the best compromise.

**Comment by Cloudmersive** [comment]
This is still broken - can we re-open this issue and fix the problem?  The current release does NOT work.

**Comment by kdavis-mozilla** [comment]
@Cloudmersive  Other got this to work[[1](https://github.com/mozilla/DeepSpeech/issues/1407#issuecomment-396575270)] so I'd guess you're doing something different to what they are doing. Could you describe what you've done?

**Comment by Hafsa26** [comment]
I am having the same error.

terminate called after throwing an instance of 'lm::FormatLoadException'
  what():  native_client/kenlm/lm/binary_format.cc:131 in void lm::ngram::MatchCheck(lm::ngram::ModelType, unsigned int, const lm::ngram::Parameters&) threw FormatLoadException.
The binary file was built for probing hash tables but the inference code is trying to load trie with quantization and array-compressed pointers
Aborted (core dumped)

Deepspeech current Master 0.2.1.Alpha 0. 
I have created the Urdu language, language model in binary formart. 
Using given link : https://yidatao.github.io/2017-05-31/kenlm-ngram/
next, I want to generate trie. I have native client in deepspeech current master.

Give me solution kindly. or identify my mistake. 
Thank you!

**Comment by Hafsa26** [comment]
I followed given link to generate trie but got errors. 
https://discourse.mozilla.org/t/tutorial-how-i-trained-a-specific-french-model-to-control-my-robot/22830

**Comment by Hafsa26** [comment]
> @yoann1995
> 
> ```
> Usage: ./build_binary [-u log10_unknown_probability] [-s] [-i] [-w mmap|after] [-p probing_multiplier] [-T trie_temporary] [-S trie_building_mem] [-q bits] [-b bits] [-a bits] [type] input.arpa [output.mmap]
> ```
> 
> So make sure `[type]` is `trie` and not `probing`, I think this is what you need.

does it to generate trie? 
It would need trained model which I didn't have yet. 
I believe that generating trie is prior process than training. what to put in output.mmp or does it just like that ?

**Comment by lissyx** [comment]
> Give me solution kindly. or identify my mistake.

Read the error and try to understand it ; it's telling you exactly where the problem lies:
>  The binary file was built for probing hash tables but the inference code is trying to load trie with quantization and array-compressed pointers

You are using mismatching things, but since you don't document what you do, how you do, then we cannot help you.

> Deepspeech current Master 0.2.1.Alpha 0.
> I followed given link to generate trie but got errors.

Help yourself and others, share more details ...

**Comment by lissyx** [comment]
> > We've very recently switched the format of our language model and tooling on master. It looks like you're using the pre-built generate_trie binary from the 0.1.1 release. You can fix this by using consistent versions of the various tools, including the training code. If you want to use the 0.1.1 binaries, work from the v0.1.1 tag, not the master branch. Alternatively you can work from master and build things yourself.
> > Hope this helps.
> 
> I am using DeepSpeech master,version 0.2.1 Alpha 0. But I am facing the same challenge.

You are facing the same challenge, but have not yet provided us with more details so that we can help you. If you cannot be more constructive, I cannot see how we can help you.

**Comment by lissyx** [comment]
So @Hafsa26 when do you think you can share us more details on your environment and what you do so we can start helping you ?

**Comment by lissyx** [comment]
>  Using given link : https://yidatao.github.io/2017-05-31/kenlm-ngram/

So @Hafsa26 that's not documentation we wrote, it's unlikely to be matching our usecase. However, if you read the documentation, we explicitely explain and instruct you how to do it: https://github.com/mozilla/DeepSpeech/blob/v0.2.1-alpha.0/data/lm/README.md

**Comment by Hafsa26** [comment]
I am using Deepspeech current Master 0.2.1.Alpha 0 and installed all its requirements. My installations worked finely with Common voice data sets. I did training as well of that datatset. 
Now, I am working on Urdu Language dataset cointaing 708 sentences. 
I prepared different directory of test, train and dev as well as .csv files of these audio files along their transcription/ the standard way of preparing data for deepspeech. 
I have created the Urdu language, language model in binary format.
Using given link : https://yidatao.github.io/2017-05-31/kenlm-ngram/
next, I want to generate trie. I have native client in deepspeech current master.
I was getting an error 
****terminate called after throwing an instance of 'lm::FormatLoadException'
  what():  native_client/kenlm/lm/binary_format.cc:131 in void lm::ngram::MatchCheck(lm::ngram::ModelType, unsigned int, const lm::ngram::Parameters&) threw FormatLoadException.
The binary file was built for probing hash tables but the inference code is trying to load trie with quantization and array-compressed pointers'
Aborted (core dumped)****
I removed this error by changing binary file into 
./build/bin/build_binary trie -a 64 -q 8 -b 7  urdu_3gram.arpa lm.binary 
using https://kheafield.com/code/kenlm/structures/ 
Now, The build lm.binary has quantization and pointer compression. 
But now, I am getting an error of
**Invalid label ﻿
Aborted (core dumped)**

Kindly help. or identify my mistake.
Thank you!

**Comment by Hafsa26** [comment]
My all other files of urdu data are also prepared and I placed them in required folders. 
Help me in it. 
Thank you!

**Comment by Hafsa26** [comment]
A text trie file  is made in the result of below command only 1 is written in it.
/home/rc//Downloads/DeepSpeech-master/native_client/generate_trie /home/rc/Downloads/DeepSpeech-master/data/alphabet.txt /home/rc/Downloads/DeepSpeech-master/data/lm/lm.binary /home/rc/Downloads/DeepSpeech-master/data/vocab.txt /home/rc/Downloads/DeepSpeech-master/data/trie
@lissyx

**Comment by Hafsa26** [comment]
Let me know if something is missing in order to identify the error. I will give more details about that then. Thank you! @lissyx

**Comment by lissyx** [comment]
>  **Invalid label ﻿**

This usually means there's some missing character in the alphabet.

**Comment by Hafsa26** [comment]
@lissyx  Okay. I'm gonna check it. Thanks!

**Comment by Hafsa26** [comment]
Am I on the right track to get the trie file ?

**Comment by lissyx** [comment]
> Am I on the right track to get the trie file ?

No, but that's because you don't want to read the documenation. You are passing too many arguments to `generate_trie` ...

**Comment by lissyx** [comment]
>  Using given link : https://yidatao.github.io/2017-05-31/kenlm-ngram/

I've already said that this is not the proper documentation to follow. If you insist on using unrelated documentation when we have the proper one in the tree under `data/lm/README.md`, we cannot continue to help you.

**Comment by Hafsa26** [comment]
I read that. 
I got from the code that trie file is quantized as well as has pointer compression; and then trie is build.

**Comment by Hafsa26** [comment]
Sorry the binary file to be built is quantized and had pointer compression. Thats what I did. 
Previously, my lm.binary has probe hash tables, now I changed it to trie along quantization and pointer compression.
